src/runtime/virtcontainers/fs_share_linux.go

```diff
@@ -0,0 +1,368 @@
+		return fmt.Errorf("Creating sandbox shared mount directory: %v: %w", sandboxMountDir, err)
+	}
+	var mountedList []string
+	defer func() {
+		if err != nil {
+			for _, mnt := range mountedList {
+				if derr := unmountNoFollow(mnt); derr != nil {
+					f.Logger().WithError(derr).Errorf("Cleanup: couldn't unmount %s", mnt)
+				}
+			}
+			if derr := os.RemoveAll(sandboxMountDir); derr != nil {
+				f.Logger().WithError(derr).Errorf("Cleanup: failed to remove %s", sandboxMountDir)
+			}
+
+		}
+	}()
+
+	for _, m := range f.sandbox.config.SandboxBindMounts {
+		mountDest := filepath.Join(sandboxMountDir, filepath.Base(m))
+		// bind-mount each sandbox mount that's defined into the sandbox mounts dir
+		if err := bindMount(ctx, m, mountDest, true, "private"); err != nil {
+			return fmt.Errorf("Mounting sandbox directory: %v to %v: %w", m, mountDest, err)
+		}
+		mountedList = append(mountedList, mountDest)
+
+		mountDest = filepath.Join(sandboxShareDir, filepath.Base(m))
+		if err := remountRo(ctx, mountDest); err != nil {
+			return fmt.Errorf("remount sandbox directory: %v to %v: %w", m, mountDest, err)
+		}
+	}
+
+	return nil
+}
+
+func (f *FilesystemShare) cleanupBindMounts(ctx context.Context) error {
+	if f.sandbox.config == nil || len(f.sandbox.config.SandboxBindMounts) == 0 {
+		return nil
+	}
+
+	var retErr error
+	bindmountShareDir := filepath.Join(getMountPath(f.sandbox.ID()), sandboxMountsDir)
+	for _, m := range f.sandbox.config.SandboxBindMounts {
+		mountPath := filepath.Join(bindmountShareDir, filepath.Base(m))
+		if err := unmountNoFollow(mountPath); err != nil {
+			if retErr == nil {
+				retErr = err
+			}
+			f.Logger().WithError(err).Errorf("Failed to unmount sandbox bindmount: %v", mountPath)
+		}
+	}
+	if err := os.RemoveAll(bindmountShareDir); err != nil {
+		if retErr == nil {
+			retErr = err
+		}
+		f.Logger().WithError(err).Errorf("Failed to remove sandbox bindmount directory: %s", bindmountShareDir)
+	}
+
+	return retErr
+}
+
+func (f *FilesystemShare) Prepare(ctx context.Context) error {
+	span, ctx := katatrace.Trace(ctx, f.Logger(), "prepare", fsShareTracingTags)
+	defer span.End()
+
+	var err error
+
+	// create shared path structure
+	sharePath := GetSharePath(f.sandbox.ID())
+	mountPath := getMountPath(f.sandbox.ID())
+	if err := os.MkdirAll(sharePath, sharedDirMode); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(mountPath, DirMode); err != nil {
+		return err
+	}
+
+	// slave mount so that future mountpoints under mountPath are shown in sharePath as well
+	if err := bindMount(ctx, mountPath, sharePath, true, "slave"); err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			if umountErr := unmountNoFollow(sharePath); umountErr != nil {
+				f.Logger().WithError(umountErr).Errorf("failed to unmount vm share path %s", sharePath)
+			}
+		}
+	}()
+
+	// Setup sandbox bindmounts, if specified.
+	if err = f.prepareBindMounts(ctx); err != nil {
```

Comment:
I understand that this is an artifact of how it is written today in `kata_agent`, but I'm not sure how I feel about the sandbox bind mounts being part of the shared filesystem setup in `Preparation`.
 
So, I suppose the alternative would be to make it explicit as part of the fs_share interface, like ShareSandboxFile, which... may be worse than just having it part of Prepare/Cleanup.

(ie, this one is unique from ShareFile, since it isn't dealing with a mount tied to a container).

Curious if you have thoughts on this.

Reply:
> I understand that this is an artifact of how it is written today in `kata_agent`, but I'm not sure how I feel about the sandbox bind mounts being part of the shared filesystem setup in `Preparation`.
> 
> So, I suppose the alternative would be to make it explicit as part of the fs_share interface, like ShareSandboxFile, which... may be worse than just having it part of Prepare/Cleanup.
> 
> (ie, this one is unique from ShareFile, since it isn't dealing with a mount tied to a container).
> 
> Curious if you have thoughts on this.

I took this approach and decided not to explicitly expose a Sandbox file sharing API because we do it implicitly in today's current code. So having an interface API for that  would mean we'd unconditionally call it for every sandbox creation. Basically I was trying to keep the interface as short as I could, we have many examples in the current code base where we've split our interfaces into methods that we call once, unconditionally, that imho should not be part of the API itself.

But that's not a super strong opinion, almost a matter of taste I guess.